File: gateaway/src/limiter/slidingWindow.js
/**
 * Sliding Window Rate Limiter — Lua-backed atomic implementation
 *
 * WHY THIS EXISTS — the fixed window problem:
 *
 *   Fixed window (60s, max 5):
 *   |─── window 1 ───|─── window 2 ───|
 *                req1 req2 req3 req4 req5 | req1 req2 req3 req4 req5
 *                                  ↑ boundary
 *   10 requests in ~2 seconds across the boundary — both windows allow it.
 *   That's a burst the fixed window can't see.
 *
 * HOW SLIDING WINDOW FIXES IT:
 *
 *   Instead of a fixed bucket, keep a rolling log of timestamps.
 *   "How many requests in the last 60 seconds FROM NOW?"
 *   The window moves with every request — no boundary exploit.
 *
 * WHY LUA — the distributed race condition:
 *
 *   Without atomicity, two gateway nodes can race:
 *
 *     Gateway A: ZCARD key → 4  (under limit, hasn't written yet)
 *     Gateway B: ZCARD key → 4  (under limit, hasn't written yet)
 *     Gateway A: ZADD key  → count becomes 5  ✅ allowed
 *     Gateway B: ZADD key  → count becomes 6  ✅ allowed ← WRONG, should be blocked
 *
 *   The Lua script runs as a single atomic Redis command — no other client
 *   can execute between ZCARD and ZADD. This is how production systems
 *   handle distributed rate limiting correctly.
 *
 * REDIS DATA STRUCTURE — Sorted Set:
 *
 *   Key:   rl:sw:ip:route:auth_login:127.0.0.1
 *   Score: timestamp in ms (e.g. 1741600123456)
 *   Value: unique request ID (timestamp:random to handle same-ms requests)
 */

import { redis } from "../redis/client.js";

export async function consumeSlidingWindow({ key, windowMs, max }) {
  const now         = Date.now();
  const windowStart = now - windowMs;

  // Unique member — timestamp + random suffix handles burst requests at same ms
  const member = `${now}:${Math.random().toString(36).slice(2, 8)}`;

  // Single atomic Lua call — replaces the two-pipeline approach.
  // Returns: [allowed, count, remaining, oldestScore, countBefore]
  const [allowed, count, remaining, oldestScore] =
    await redis.slidingWindowConsume(
      key,          // KEYS[1]
      now,          // ARGV[1]
      windowStart,  // ARGV[2]
      windowMs,     // ARGV[3]
      max,          // ARGV[4]
      member,       // ARGV[5]
    );

  const retryAfterMs = windowMs - (now - oldestScore);

  return {
    allowed:      allowed === 1,
    limit:        max,
    current:      count,
    remaining:    remaining,
    retryAfterMs: retryAfterMs > 0 ? retryAfterMs : windowMs,
  };
}

File: gateaway/src/limiter/scripts/slidingWindow.lua
--[[
  Atomic Sliding Window Rate Limiter
  ===================================
  Runs entirely inside Redis — no round-trips between gateway and Redis between
  steps. This is the key difference from the two-pipeline JS approach: the
  entire check-and-consume is one atomic operation.

  Why this matters for distributed gateways:
  -------------------------------------------
  Without atomicity, two gateway nodes can race:

    Gateway A:  ZCARD key  → count = 4  (under limit)
    Gateway B:  ZCARD key  → count = 4  (under limit — A hasn't written yet)
    Gateway A:  ZADD key   → count = 5  ✅ allowed
    Gateway B:  ZADD key   → count = 6  ✅ allowed — BUT SHOULD HAVE BEEN BLOCKED

  With this Lua script, Redis executes everything as a single command.
  No other client can interleave between the ZCARD and the ZADD.

  Arguments (ARGV):
    ARGV[1]  now           — current timestamp in ms
    ARGV[2]  windowStart   — now - windowMs (oldest allowed timestamp)
    ARGV[3]  windowMs      — window size in ms (for PEXPIRE)
    ARGV[4]  max           — maximum requests allowed in the window
    ARGV[5]  member        — unique member string for this request

  Keys (KEYS):
    KEYS[1]  — the sorted set key for this client+route

  Returns a flat table (becomes a Redis array reply):
    [1]  allowed        (1 = yes, 0 = no)
    [2]  count          (requests in window AFTER this one, if allowed)
    [3]  remaining      (max - count, floored at 0)
    [4]  oldestScore    (timestamp ms of oldest entry, for retryAfter calc)
    [5]  countBefore    (requests in window BEFORE this one)
--]]

local key         = KEYS[1]
local now         = tonumber(ARGV[1])
local windowStart = tonumber(ARGV[2])
local windowMs    = tonumber(ARGV[3])
local max         = tonumber(ARGV[4])
local member      = ARGV[5]

-- Step 1: Remove all entries outside the rolling window
redis.call("ZREMRANGEBYSCORE", key, 0, windowStart)

-- Step 2: Count how many requests are in the window RIGHT NOW (before adding)
local countBefore = redis.call("ZCARD", key)

-- Step 3: Deny immediately if already at limit — don't consume a slot
if countBefore >= max then
  -- Find the oldest entry to calculate retryAfter
  local oldest = redis.call("ZRANGE", key, 0, 0, "WITHSCORES")
  local oldestScore = now
  if #oldest >= 2 then
    oldestScore = tonumber(oldest[2])
  end
  return { 0, countBefore, 0, oldestScore, countBefore }
end

-- Step 4: Allowed — add this request's timestamp and reset TTL
redis.call("ZADD", key, now, member)
redis.call("PEXPIRE", key, windowMs)

local count = countBefore + 1
local remaining = max - count

-- Find oldest entry for retryAfter (now includes the entry we just added)
local oldest = redis.call("ZRANGE", key, 0, 0, "WITHSCORES")
local oldestScore = now
if #oldest >= 2 then
  oldestScore = tonumber(oldest[2])
end

return { 1, count, remaining, oldestScore, countBefore }
